pkg/R/types.R

```diff
@@ -0,0 +1,47 @@
+# Utility functions for handling Spark DataTypes.
+
+# Handler for StructType
+structType <- function(st) {
+  obj <- structure(new.env(parent = emptyenv()), class = "structType")
+  obj$jobj <- st
+  obj$fields <- lapply(SparkR:::callJMethod(st, "fields"), structField)
```

Comment:
I think @shivaram's proposal is worth a shot (@cafreeman not sure if you have tried that). Also PipelinedRDD's getJRDD pattern is an example of how memoization could be done.